### ISSUE TITLE
Bug: [explicit-function-return-type] false invalid case when using allowTypedFunctionExpressions option

### ISSUE THREAD

**Issue by kaykdm** [typescript-eslint/typescript-eslint#9171]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.4.5&fileType=.ts&code=C4TwDgpgBAYgrgOwMYBVzQLxQBQEooYB8UAzsAE4CWCA5gNwBQTANhMFAIbnkD2A7jAQBGAFyxEqdIy68BwgjnxEoAcmAQyKxgyQ8EZTt36CATGPjI0kBXgLE1G4FqA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6RAD2MbMroDNYTMvkoB7JrWiJ8saJKKl0YANrhsORNGhjokADTqNWDdkgBDePDEB3AColEAEwBiQkeKYBRLtOTJPVAx8OEQjbABfcIBddSiIoA&tsconfig=&tokens=false

### Repro Code

```TypeScript
type FuncType = () => string;

let arrowFn1: FuncType;
arrowFn1 = () => 'test'; // Error

const arrowFn2: FuncType = () => 'test'; // No Error as expected

```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/explicit-function-return-type": ["error", {"allowTypedFunctionExpressions": true}],
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

No Error

### Actual Result

Error on line 5

### Additional Info

If the `allowTypedFunctionExpressions ` option is true and we re-assign value to typed variable, we get an error while using const with type is valid. (see: Repro Code)

I'm not sure this is a bug or not, but it seems like it is. However this is covered in invalid test cases, so this is not a bug?
https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/tests/rules/explicit-function-return-type.test.ts#L1063-L1084

**Comment by Josh-Cena** [comment]
I think this should be allowed. The test you quoted is typing a higher-order function. cc @typescript-eslint/triage-team

**Comment by JoshuaKGoldberg** [comment]
A variable might not be "typed":

```ts
let evolvingAny;
evolvingAny = (param) => param;
```

So we'd now have to use scope analysis to determine whether the variable has a type? I'm ok with this, with the note that this rule is one that I don't think we often recommend.

**Comment by kirkwaiblinger** [comment]
+1 I think that if the variable declaration is easily seen with the scope analyzer it makes sense to consider it typed.

**Comment by bradzacher** [comment]
The big reason we didn't do this is because it requires scope analysis. It adds a substantial amount of complexity for what is ultimately little gain.
The rule is already very complex as it is - so the extra complexity isn't worth it, IMO.

Additionally consider code like this:

``ts
let x: () => string;

//
// 100 lines
//

x = () => {...};
```

It can be very hard for a human to go and find a variables original declaration to then figure out its annotated type.

Esp in contexts without IDE tooling (eg GH reviews) where you can't just "go to definition". One of the major reasons behind explicit return types is to declare contracts close to the function itself. But with this you can have an indeterminate number of lines between the two.

I'm personally a -1 on this for the above reasons.

**Comment by kirkwaiblinger** [comment]
I guess this is kinda the same as https://github.com/typescript-eslint/typescript-eslint/issues/8918. 

I see two possible principles behind using this rule.
1. prevent writing a function whose return type is unconstrained and determined by its implementation
1. prevent writing a function with a return type that cannot be determined by a human reader without LSP support (this is a stricter condition than `1.`)

IMO, under principle `1.`, this is clearly a bug, and under principle `2.` there is a pretty strong argument that this is working as intended.

For me, I personally only care about `1.` in my usage of the rule. 

The [rule docs page](https://typescript-eslint.io/rules/explicit-function-return-type/) aligns more with `2`.

@bradzacher I'm curious for your assessment, but based on https://github.com/typescript-eslint/typescript-eslint/issues/9171#issuecomment-2144200226 and https://github.com/typescript-eslint/typescript-eslint/issues/8918#issuecomment-2056908629, I would think your thoughts align better with `2.` as well.

***

What I would say, though, is I think the option `allowTypedFunctionExpressions` itself feels like an option in service of principle `1.` and an explicit opt-out of principle `2.`, since `let f: NamedType = () => {}` also can't be resolved by an unassisted human. So, for the moment, I remain a +1 to considering this a bug, since I think `1.` is the right standard to apply here. But, my opinion will be changed if convinced that principle `2.` is the standard to apply.

**Comment by bradzacher** [comment]
I don't think it's an either-or scenario.
Both principles hold true.

You want functions to have strict contracts because it tightens the boundaries between code and reduces actions at a distance caused by unintended changes in the return type.

But you also want to declare strict contracts because it makes reading and reviewing code so much easier. Being able to read a function at a glance as `arguments => return` is invaluable in scanning and understanding code. In an IDE this can ofc be supplanted with intellisense et al - though that does require a manual action to see. Inlay hints also help though from my experience they're rarely turned on in a way that covers all the gaps (if they're turned on at all).

In many usecases (eg browser / github reviews / screenshots (ew)) IDE features cannot be used so the strict contracts provide a lot of clarity and greatly assist readability.

You're not wrong that `allowTypedFunctionExpressions` is a weird fit for the latter justification. It's ultimately a DevX concession. It's possible to work without it - but you end up duplicating annotations and it creates resentment towards the style. Ultimately an annotated name is (generally) greppable and easy to find so you don't lose a massive amount.

**Comment by kirkwaiblinger** [comment]
> I don't think it's an either-or scenario. Both principles hold true.

Right, the set of report locations of `1.` is completely contained within `2.`, so I take `2.` to imply `1.`, though I suppose, philosophically, those could be separate intentions. Anyways, semantics! 🙂 
 
> You want functions to have strict contracts because it tightens the boundaries between code and reduces actions at a distance caused by unintended changes in the return type.
> 
> But you also want to declare strict contracts because it makes reading and reviewing code so much easier. Being able to read a function at a glance as `arguments => return` is invaluable in scanning and understanding code. In an IDE this can ofc be supplanted with intellisense et al - though that does require a manual action to see. Inlay hints also help though from my experience they're rarely turned on in a way that covers all the gaps (if they're turned on at all).
> 
> In many usecases (eg browser / github reviews / screenshots (ew)) IDE features cannot be used so the strict contracts provide a lot of clarity and greatly assist readability.
> 
> You're not wrong that `allowTypedFunctionExpressions` is a weird fit for the latter justification. It's ultimately a DevX concession. It's possible to work without it - but you end up duplicating annotations and it creates resentment towards the style. Ultimately an annotated name is (generally) greppable and easy to find so you don't lose a massive amount.

Yeah, basically I would interpret this as a compelling `2.-with-devx-concession` take. And I would agree that the current behavior is the correct behavior under that interpretation, since it's the minimal concession to satisfy the devx complaint. So, my opinion has been changed!.

I would be sympathetic to the creation of a mode for flagging for principle `1.` concerns only, in which case the code in this issue would be allowed, but that'd be a separate feature proposal rather than a bug.

**Comment by JoshuaKGoldberg** [comment]
@bradzacher @kirkwaiblinger following up: is there consensus here? If I'm reading right, was it to make a new option for the rule that allows for cases like this and #8918?

**Comment by bradzacher** [comment]
Personally - I think we should reject this.
The rule has worked this way for a long, long time and in 3 months this issue has garnered 0 reactions and 0 other community comments.

I just don't think that the majority of the community (a) uses this sort of "define variable then assign function expression" pattern (b) cares about the additional annotation required.

So yeah -1 on this.

**Comment by kirkwaiblinger** [comment]
Yeah, I'm good with rejecting. Based on the way the rule currently is intended to work, it's a reasonable behavior. Creating an option seems like medium-or-more effort and increase in rule complexity for low value.
